Add support for label parameter on docker image creation

jmewes opened the pull request.

jmewes commented:
This pull request is based in the pull request with the version bump for `docker-java`: https://github.com/bmuschko/gradle-docker-plugin/pull/331

bmuschko commented on src/main/groovy/com/bmuschko/gradle/docker/DockerRemoteApiPlugin.groovy:
```diff
@@ -47,14 +47,14 @@ class DockerRemoteApiPlugin implements Plugin<Project> {
                 project.repositories.addAll(project.buildscript.repositories.collect())
             }
         }
-        
```
What has changed here? Spaces to tab?

jmewes replied:
No, my IDE automatically removes trailing spaces. I can revert it in the pull request for the version update if it disturbs.

bmuschko commented on src/main/groovy/com/bmuschko/gradle/docker/DockerRemoteApiPlugin.groovy:
```diff
@@ -47,14 +47,14 @@ class DockerRemoteApiPlugin implements Plugin<Project> {
             dependencies.add(project.dependencies.create('org.slf4j:slf4j-simple:1.7.5'))
             dependencies.add(project.dependencies.create('cglib:cglib:3.2.0'))
         }
-        
```
What has changed here? Spaces to tab?

jmewes replied:
Same as above.

bmuschko commented on src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy:
```diff
@@ -61,6 +61,10 @@ class DockerBuildImage extends AbstractDockerRemoteApiTask implements RegistryCr
 
     @Input
     @Optional
+    Map<String, String> labels
```
Should have an empty `Map` as default value.

bmuschko commented on src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy:
```diff
@@ -111,6 +115,10 @@ class DockerBuildImage extends AbstractDockerRemoteApiTask implements RegistryCr
             buildImageCmd.withPull(getPull())
         }
 
+        if(getLabels()) {
```
Coding style right now is to have a space after `if` and the opening brace. Sorry for being anal.

cdancy replied:
Created [ISSUE](https://github.com/bmuschko/gradle-docker-plugin/issues/334) to implement checkstyle support so that users will be able to catch these locally and/or as part of CI.

jmewes replied:
@bmuschko It's fine. Actually caring about all details is a good thing.

If @bklein1 would not have written this in the train he would probably had seen it by himself :)

bmuschko commented on src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImageFunctionalTest.groovy:
```diff
@@ -52,6 +51,16 @@ class DockerBuildImageFunctionalTest extends AbstractFunctionalTest {
         result.output.contains("label1:test1, label2:test2")
     }
 
+    def "can build image with labels without using build-args"() {
```
We should merge the test cases "can build image with build variables" and "can build image with labels without using build-args". Both execute the same task and assert the same output. Let's use a single test method with a `where` clause. Annotate the method with `@Unroll`.

bmuschko commented on src/main/groovy/com/bmuschko/gradle/docker/DockerRemoteApiPlugin.groovy:
```diff
@@ -29,7 +29,7 @@ import org.gradle.api.file.FileCollection
 class DockerRemoteApiPlugin implements Plugin<Project> {
     public static final String DOCKER_JAVA_CONFIGURATION_NAME = 'dockerJava'
-    public static final String DOCKER_JAVA_DEFAULT_VERSION = '3.0.6'
+    public static final String DOCKER_JAVA_DEFAULT_VERSION = '3.0.7'
```
Is a new version of Docker Java required for supporting labels? I'd prefer a new PR for the change of the Docker Java library if it is not required.

bmuschko replied:
I see this has already been done in https://github.com/bmuschko/gradle-docker-plugin/pull/331. Please remove it from this PR.

jmewes replied:
Yes, it is required. AFAIK, with `3.0.6` it is only possible to declare label via build-args. [Here](https://github.com/docker-java/docker-java/commit/305b73a9763f8d47bdc9698ecfaa9c844a6e087f#diff-d7f48445f5b749f480d3c3d0eb78f6ef) it has been enabled to declare the labels directly.

bmuschko commented:
@cdancy Can you follow up on the PR? I already did a review. Seems like the functionality is highly valuable to some users of the plugin. Thanks!

cdancy commented:
Thanks @bmuschko! 

@jmewes please address the comments and we can move forward from there. The main thing from my end is to remove the "docker-java upgrade" from this PR as it's being addressed elsewhere (by yourself no less ;)).

cdancy commented:
@jmewes you don't need to remove the "space" changes. Better to remove them anyway IMO as it creates cleaner looking code. Checkstyle support _should_ check for things like this anyway and I suspect once implemented there will be a fair amount of tedious work to do to get the project into pristine shape.

jmewes commented:
@bmuschko, @cdancy Thank you for your fast response! I have addressed all the review remarks and updated the pull request.

I am a bit surprised that the build is still green after pulling out 1855d1a5622d6bdb713e47e0d0f4ad36e5b67a6a from this pull request. When I started working on this issue, I had an error in our downstream project, complaining that the method `BuildImageCmd#withLabels` cannot be found. After I did some hacks in my Gradle caches to use the then [unreleased code mentioned above](https://github.com/docker-java/docker-java/commit/305b73a9763f8d47bdc9698ecfaa9c844a6e087f#diff-d7f48445f5b749f480d3c3d0eb78f6ef) it worked out.

cdancy commented:
LGTM. Integration tests pass as expected. Thanks for the PR @jmewes !

jmewes commented:
@cdancy , @bmuschko : Thank you once again for the collaboration and quick response!
